### PR TITLE
[WIP] scans program rule condition to filter OrgUnitGroups

### PR DESCRIPTION
**Pull request by zubaira**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java [hunk 1]
```diff
@@ -57,6 +59,11 @@
 
     private static final String USER = "USER";
 
+    private static final String REGEX = "d2:inOrgUnitGroup\\( *(([\\d/\\*\\+\\-%\\. ]+)|" +
+        "( *'[^']*'))*( *, *(([\\d/\\*\\+\\-%\\. ]+)|'[^']*'))* *\\)";
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java]: What is this regexp doing? We have a UID-regex in ExpressionService which can be reused somehow if we are looking for UIDs here.

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java [hunk 2]
```diff
@@ -169,20 +176,43 @@
...
 
-        List<OrganisationUnitGroup> groups = organisationUnitGroupService.getAllOrganisationUnitGroups();
+        if( !orgUnitGroups.isEmpty() )
+        {
+            supplementaryData = orgUnitGroups.stream().collect( Collectors.toMap( g -> g,  g -> organisationUnitGroupService.getOrganisationUnitGroup( g ).getMembers()
+                .stream().map( OrganisationUnit::getUid ).collect(Collectors.toList() ) ) );
+        }
 
-        groups.stream().forEach( group -> supplementaryData.put( group.getUid(), group.getMembers().stream().map( OrganisationUnit::getUid ).collect( Collectors.toList() ) ) );
+        if( currentUserService.getCurrentUser() != null )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java]: Code style, space after if

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java [hunk 3]
```diff
@@ -169,20 +176,43 @@
...
-            supplementaryData.put( USER, currentUserService.getCurrentUser().getUserCredentials().getUserAuthorityGroups().stream().map( UserAuthorityGroup::getUid ).collect( Collectors.toList() ) );
+            Rule rule = programRuleEntityMapperService.toMappedProgramRule( programRule );
+
+            if( rule != null )
+            {
+                rules.add( rule );
+
+                Matcher matcher = PATTERN.matcher( StringUtils.defaultIfBlank( programRule.getCondition(), "" ) );
+
+                while( matcher.find() )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java]: Codestyle, space

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/hibernate/HibernateProgramRuleStore.java [hunk 4]
```diff
@@ -62,7 +65,20 @@ public ProgramRule getByName( String name, Program program )
...
+    {
+        Query q = getQuery("FROM ProgramRule pr " +
+            "JOIN FETCH pr.programRuleActions pra  WHERE pr.program = :programId AND pra.programRuleActionType IN (:implementableTypes)" );
+
+        q.setParameter( "programId", program );
+        q.setParameter( "implementableTypes", types );
+
+        return (List<ProgramRule>) q.getResultList();
+    }
+
```
Comment (vietnguyen) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/hibernate/HibernateProgramRuleStore.java]: We can remove @SuppressWarnings("unchecked") with this 
```suggestion
return getQuery("FROM ProgramRule pr " +
             "JOIN FETCH pr.programRuleActions pra  WHERE pr.program = :programId AND pra.programRuleActionType IN (:implementableTypes)" )
         .setParameter( "programId", program );
         .setParameter( "implementableTypes", types )
         .getResultList();
```

**Pull request merged**